scan-payment/src/main/java/com/getbouncer/scan/payment/analyzer/NameDetectAnalyzer.kt

```diff
@@ -0,0 +1,268 @@
+        val predictions: MutableList<CharPredictionWithBox> = ArrayList<CharPredictionWithBox>()
+        var begin = 0
+
+        while (begin < width - height) {
+            val firstLetterBitmap = Bitmap.createBitmap(nameBitmap, begin, 0, height, height)
+
+            predictions.add(CharPredictionWithBox(
+                alphabetDetect.analyze(AlphabetDetect.Input(firstLetterBitmap), Unit),
+                RectF(begin.toFloat(), 0F, height.toFloat(), height.toFloat())
+            ))
+
+            begin += height / NUM_PREDICTION_STRIDES
+        }
+        val boxes = Array(predictions.size) { FloatArray(4) }
+        val probabilities = FloatArray(predictions.size)
+        for (i in predictions.indices) {
+            boxes[i] = predictions[i].getPercentBoxArray(
+                bitmapForObjectDetection.width,
+                bitmapForObjectDetection.height
+            )
+            probabilities[i] = predictions[i].characterPrediction.confidence
+        }
+        val indices: List<Int> = hardNonMaximumSuppression(
+            boxes, probabilities,
+            NMS_THRESHOLD,
+            0
+        )
+        return processNMSResults(indices.sorted(), predictions)
+    }
+
+    /**
+     * Processes each cluster of letters from NMS, doing a simple voting algorithm and
+     * tie-breaking with confidence
+     */
+    private fun processNMSCluster(charClusters: List<AlphabetDetect.Prediction>): Char {
+        var candidateLetter = 0.toChar()
+        var candidateLetterConfidence = 0f
+        var candidateConsecutiveCount = 0
+        var currentConsecutiveCount = 0
+        var currentLetterMaxConfidence = 0f
+        var lastSeenLetter = 0.toChar()
+        for (characterPrediction in charClusters) {
+            if (lastSeenLetter == characterPrediction.character) {
+                currentConsecutiveCount += 1
+                currentLetterMaxConfidence =
+                    Math.max(currentLetterMaxConfidence, characterPrediction.confidence)
+            } else {
+                currentConsecutiveCount = 1
+                currentLetterMaxConfidence = characterPrediction.confidence
+                lastSeenLetter = characterPrediction.character
+            }
+            if (currentConsecutiveCount == candidateConsecutiveCount && currentLetterMaxConfidence > candidateLetterConfidence ||
+                currentConsecutiveCount > candidateConsecutiveCount
+            ) {
+                candidateLetterConfidence = currentLetterMaxConfidence
+                candidateLetter = characterPrediction.character
+                candidateConsecutiveCount = currentConsecutiveCount
+            }
+        }
+        return if (candidateLetterConfidence > CHAR_CONFIDENCE_THRESHOLD) {
+            candidateLetter
+        } else ' '
+    }
+
+    /**
+     * Black magic to calculate the "width" (in terms of prediction index) for each space
+     */
+    private fun getSpacesWidth(spaces: List<Int>): Int {
```

Comment:
black magic is a big yikes. I suppose there's not much we can do to clarify this?